### PR TITLE
Don't scale down ASGs when placeholder nodes are discovered

### LINKED ISSUE #5829: [AWS] Unsafe decomissioning of nodes when ASGs are out of instances
**Which component are you using?**: cluster-autoscaler

**What version of the component are you using?**: cluster-autoscaler 1.25.1

Component version: chart cluster-autoscaler-9.28.0

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.9-eks-0a21954", GitCommit:"eb82cd845d007ae98d215744675dcf7ff024a5a3", GitTreeState:"clean", BuildDate:"2023-04-15T00:37:59Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
</pre></details>

**What environment is this in?**: AWS EKS

**What did you expect to happen?**: We are running EKS on managed nodes using cluster-autoscaler to dynamically scale up and down the sizes of the node groups. Unfortunately, sometimes AWS is unable to fulfil requests for scaling up (Could not launch On-Demand Instances. InsufficientInstanceCapacity). When that happens, cluster-autoscaler gets confused about the size of the ASG and sends unnecessary `SetDesiredCapacity` requests, which result in unsafe scaling down of the ASG (details below).
The expectation is that CA handles these situations transparently and does not scale down the ASG.

**What happened instead?**: We are observing the following behavior, by looking at both the CA logs and the CloudTrail logs of the requests sent. I don't know enough about the internal workings of the CA, so some of this is an assumption.
* Pods are pending, waiting for capacity in a given node group.
* CA wants to scale up this node group and sends a SetDesiredCapacity request to the ASG to increase the size (say from 3 -> 5). This happens async, so the CA fires off the request and goes back to doing different things. However it updates its internal representation of the ASG to be at capacity 5.
* ASG can’t scale up and returns that info to the CA.
* CA comes back around to the request and realizes the ASG is blocked. It updates its internal representation (back to 3)  but crucially for some reason it also updates the ASG, sending a number of SetDesiredCapacity requests, to revert the capacity back down to 3.
* In the meantime, a new instance might have successfully started, so the actual capacity of the ASG is now 4.
* So as a result to receiving the SetDesiredCapacity requests from the CA, the ASG terminates one (or more) of its running instances resulting in the unsafe scale down.

image

See the attached image that shows the sequence of events as well.

**How to reproduce it (as minimally and precisely as possible)**: Difficult to reproduce as it only happens when the AWS capacity issues occur. We are regularly observing this behavior though, I am happy to assist with more information.

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by theintz**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #5829 

#### Does this PR introduce a user-facing change?

```release-note
[AWS] ASGs are no longer scaled down when there are capacity issues within the ASG
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by theintz** [comment]
We've deployed this patch on our clusters a week ago (patched into v1.26.3) and it appears to have stopped the problems outlined in #5829. Looking forward to your feedback.

**Comment by qianlei90** [comment]
I'm not familiar with AWS's ASG, but I think one use case should be consider(correct me if I'm wrong)：

If EC2 is out of stock and ASG can not create new instances, CA will delete these nodes after MaxNodeProvisionTime(default `15m`). If the desired instance number of ASG is not modified, new instances will still be created and registry to cluster when stock is restored, which is not the expected behavior.

**Comment by jlamillan** [comment]
By removing the node(s), but not scaling down the corresponding ASG, the CA seems to be putting itself in a state where it is requesting more nodes than the cluster workload requires. 

I thought the AWS provider already [handled](https://github.com/kubernetes/autoscaler/pull/4489) this situation more or less. i.e. the scale up operations fails quickly short-circuiting the `--max-node-provision-time` if capacity errors are observed via `DescribeScalingActivities`.

File: cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go [hunk 1]
```diff
@@ -308,11 +304,9 @@ func (m *asgCache) DeleteInstances(instances []*AwsInstanceRef) error {
 
 	for _, instance := range instances {
 		// check if the instance is a placeholder - a requested instance that was never created by the node group
-		// if it is, just decrease the size of the node group, as there's no specific instance we can remove
+		// if it is, simply remove it from the cache
```
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go:307]: By removing the node(s) in the local cache, but not actually scaling down the corresponding ASG, the ongoing/pending request for additional capacity on the provider may impact what capacity is accessible subsequent requests. Basically the ongoing request is at best a capacity spoof and capacity dead-lock at worst.

Put another way, if you request a 100 instances in ASG_1, and get 25 before running into a `InsufficientInstanceCapacity`, it's probably not the right answer to leave that request on-going. It's not unreasonable to think that the ongoing request for an additional 75 nodes in ASG_1 could affect subsequent requests to provision instances of that type in the same or other ASGs.

**Comment by Shubham82** [comment]
Closing this PR, as issue #5829 is fixed by PR #6911 

Please feel free to reopen if you have any concerns.

/close

**Pull request closed without merging**